zebra-rpc/src/methods.rs

```diff
@@ -869,6 +994,22 @@ pub struct GetBlock(#[serde(with = "hex")] SerializedBlock);
 #[derive(Copy, Clone, Debug, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
 pub struct GetBestBlockHash(#[serde(with = "hex")] block::Hash);
 
+/// Response to a `z_gettreestate` RPC request.
+///
+/// Contains the hex-encoded Sapling & Orchard note commitment trees, and their
+/// corresponding [`block::Hash`], [`Height`], and block time.
+#[derive(Clone, Debug, Eq, PartialEq, serde::Serialize)]
+pub struct GetTreestate {
```

Comment:
This struct requires somes adjustments so that we [match the response format](https://zcash.github.io/rpc/z_gettreestate.html). For example, instead of a `sapling_tree` key, we need a `sapling` key with a `commitments` subkey, which in turns has `finalRoot` (the anchor) and `finalState` (the actual sapling_tree) subkeys. You can achieve that with nested structs (but maybe there's a less verbose way, if you want to look into it, but that's not required).

Also it seems we're returning the tree with some zeroes at the end, which do not show up in zcashd response. I'm not sure why, I'll try to investigate it in a bit. Here's the comparison:

Our return:

```bash
curl --data-binary '{"jsonrpc": "1.0", "id":"curltest", "method": "z_gettreestate", "params": ["1648055"] }' -H 'Content-Type: application/json' http://127.0.0.1:28232/
```

```json
{
    "hash": "00000000006c955925a3602b7b425e92d3e2c1a21b3586c477e23076f4351003",
    "height": 1648055,
    "time": 1651074546,
    "sapling_tree": "01202ac7ffd0246cede1dc9e7d4880ebfa31a0d5c907ba2a94c6a4c920a7762c6c001f0000016fbbb5ceffa468e903ab2d502eddad4d7ecbd93ada35a8c119968fcd363c423a00014761b2cc20da6c0d546919ee0bd1d15e582eff449cc0d5e8f28a33c4fa523021016b189b58d134947e9b6ffb2364dcaf7a4979b681298680ef8bfaef384d94461301c749d8b21ac8f6bc72f6d8eb4123f1baa5252c0b3534bd33fbdf8e5743fa5e1d00000110cc98733720065bb6e45470cb660d6c24c41def02683b2e57e7b75ef893645501bed24c233b57211ab90f07ea319fb53a2c2b2dcde3a9653a30df18446017f465000001232675992d1b5406f0c4a35c977f06319f3e4644ac1d165b0d0b9399f16f5d2600017eda713cd733ccc555123788692a1876f9ca292b0aa2ddf3f45ed2b47f027340000000015ec9e9b1295908beed437df4126032ca57ada8e3ebb67067cd22a73c79a840090000000000000000000000",
    "orchard_tree": "000000"
}
```

Zcashd:

```bash
zcash-cli z_gettreestate 1648055
```

```json
{
  "hash": "00000000006c955925a3602b7b425e92d3e2c1a21b3586c477e23076f4351003",
  "height": 1648055,
  "time": 1651074546,
  "sprout": {
    "commitments": {
      "finalRoot": "8a13d46e012838e2f778247e2411b08418294fe3d910bedacb79f9bb9b3c5566",
      "finalState": "015373f13240446a202f81ae3281d2f18c6bfa1e48d1cc0a745c280dc0eff46f5001c4e95cdc690dd46d53f5b72eb817ac915fe3c49e89f82140e16d3cefca4ecb0b140171baa1a5c6ba80c7b72d337a6ca6ee94d9fd6e50e40271b13a9c74e0056f8df4000001d88e185741730c8c5a8f5a961b574bf4834885a15b82ad2d7154a4bfc0e5773e0109eca6fe93c2f849d55b0cc775d59f856ef3e726a7eb3d12f36c1a1368a7faf201517095a7615d6998030b3cc89ac91ac3255ec2da9e66389cff5cbca78701e73f0001f779c93f8bae99232c50676ee35b5ee0b16bb291b3f0e3ebea59b195c724f25a0118a42f7941cd9d5f1459d0ae25ebf4528402d7db7b3e6f56aaaf1d8b49a8ca3500000196295a1e8d018c1c59f098e69350c7a206b850c828710fbaeb86d3ef72da3bb6000001806212ff57170f6904d462d24bc5df6112269a0b65defffbd4b78ff9a7859b9d0183a03915c566ab12ba59b130bab08a25b932ce204bf1774fc9e9267e5130f19101465dbb7f98de7f19e22f37c537660e9447fc27e1839238dc75548b047c41d1be01c14d7b1dfae9c25c284218d74c43621b795246a5fb3c238c2ca9d7f91d5aa86000011b1fc3241793d87854115ec7190ea06c64b76faa971c77ccc6316ec716f90f77"
    }
  },
  "sapling": {
    "commitments": {
      "finalRoot": "2a34d72c27cff5e4a8433ed542c75f466e3457f47070064caca46a464cce1b85",
      "finalState": "01202ac7ffd0246cede1dc9e7d4880ebfa31a0d5c907ba2a94c6a4c920a7762c6c00140000016fbbb5ceffa468e903ab2d502eddad4d7ecbd93ada35a8c119968fcd363c423a00014761b2cc20da6c0d546919ee0bd1d15e582eff449cc0d5e8f28a33c4fa523021016b189b58d134947e9b6ffb2364dcaf7a4979b681298680ef8bfaef384d94461301c749d8b21ac8f6bc72f6d8eb4123f1baa5252c0b3534bd33fbdf8e5743fa5e1d00000110cc98733720065bb6e45470cb660d6c24c41def02683b2e57e7b75ef893645501bed24c233b57211ab90f07ea319fb53a2c2b2dcde3a9653a30df18446017f465000001232675992d1b5406f0c4a35c977f06319f3e4644ac1d165b0d0b9399f16f5d2600017eda713cd733ccc555123788692a1876f9ca292b0aa2ddf3f45ed2b47f027340000000015ec9e9b1295908beed437df4126032ca57ada8e3ebb67067cd22a73c79a84009"
    }
  },
  "orchard": {
    "commitments": {
      "finalRoot": "0000000000000000000000000000000000000000000000000000000000000000"
    }
  }
}
```

Reply:
(Also zcashd seems to omit empty trees, while we return their encoding. I'm not sure if this is an issue or not, it seems fine to leave as is)

Reply:
> Also it seems we're returning the tree with some zeroes at the end, which do not show up in zcashd response. I'm not sure why, I'll try to investigate it in a bit.

One slightly disturbing possibility is that they're using C strings for binary data, so trailing zeroes get automatically removed.

Another alternative is that they deliberately strip trailing zeroes.